src/com/palmergames/bukkit/towny/object/Town.java

```diff
@@ -1375,4 +1377,27 @@ public boolean collect(double amount, String reason) throws EconomyException {
 	public String getTownBoard() {
 		return getBoard();
 	}
+
+	public boolean isRuined() {
+		if(residents.size() == 0)
+			ruined= true; //If no residents, setup town for full deletion
+
+		return ruined;
+	}
+
+	public void setRuined(boolean ruined) {
+		this.ruined = ruined;
+	}
+
+	public int getRuinDurationRemainingHours() {
+		return ruinDurationRemainingHours;
+	}
+
+	public void setRuinDurationRemainingHours(int ruinDurationRemainingHours) {
+		this.ruinDurationRemainingHours = ruinDurationRemainingHours;
+	}
+
+	public void decrementRemainingRuinTimeHours() {
+		ruinDurationRemainingHours--;
+	}
```

Comment:
I'm not sure how I feel about this being put directly in the `Town` object. These methods are a specialization of a Town in a siege war situation and not general purpose. I would recommend to create a subclass of `Town` and implement the methods specific to siege war there, for clarity.

 For example:

```java
public class SiegedTown extends Town {
	
	final Town town;
	boolean ruined;
	int ruinDurationRemainingHours;
	
	public static SiegedTown from(Town town) {
		return new SiegedTown(town);
	}

	private SiegedTown(Town town) {
		super(town.getName());
		this.town = town;
	}

	public boolean isRuined() {
		if(getResidents().size() == 0)
			ruined= true; //If no residents, setup town for full deletion

		return ruined;
	}

	public void setRuined(boolean ruined) {
		this.ruined = ruined;
	}

	public int getRuinDurationRemainingHours() {
		return ruinDurationRemainingHours;
	}

	public void setRuinDurationRemainingHours(int ruinDurationRemainingHours) {
		this.ruinDurationRemainingHours = ruinDurationRemainingHours;
	}

	public void decrementRemainingRuinTimeHours() {
		ruinDurationRemainingHours--;
	}
}
```

Reply:
Hi,

- The ruin feature is generic to multiple war systems. For example it will fix the issue in event war where conquered towns can escape by delete+refound (there is an ticket for this)
- For a town to go to a ruined state, it does not need to be besieged, or involved in any sort of war. The feature can be used as per the description to destroy a town to deny it to an approaching army, or an occupier.
- The 'conquered' flag is already in the Town object, and is common to multiple war systems. The ruined flag is a similar multi-war-system feature.

Reply:
Let's first address this point:
> The 'conquered' flag is already in the Town class, and is common to multiple war systems. The ruined flag is a similar multi-system feature.

Had I been a part of towny in the first place I would've not had this field in the class.

> The ruin feature is generic to multiple war systems. For example it will fix the issue in event war where conquered towns can escape by delete+refound (there is an ticket for this)

Even though this is for a generic war system it still doesn't change the fact that "war" in general is out of the scope of the `Town` object. This is mainly because there are a servers that don't use the war system at all.

> It is even useful without a war system, if a server wants a town-upkeep-fall system which is softer than the default but harder than bankruptcy.

Yes but I really don't see how people will use this even when bankruptcy is standard. Locking a mayor from deleting/leaving the town seems more proper for war.

 But what I'm really getting at is the responsibility of objects. Towns are meant to manage/own townblocks, and functionality *purely* specific to it such as it's board, if it's open or not, and if it's public. This is the motivation for many reasons around how the town was designed, for example a bit ago we changed how accounts work for Towns and Nations. You used to interact with it like this:

```java
town.collect(100);
town.pay(1000, otherTown);
```

We refactored it so it now looks like this:
```java
town.getAccount().collect(100);
town.getAccount().pay(1000, otherTown);
```

This change was made after the realization that a `Town` or `Nation` doesn't need to be responsible for *managing* it's economics, but instead, it should only be *aware* of the state of its economics. Having the `collect` and `pay` (and some others) methods directly in the class itself it 1) made the class more bloated and 2) put too much responsibility on `Town` for things it shouldn't be responsible for. This change also makes things more clear, we're using `getAccount()` which means we're operating in the scope of an account now, and not the scope of a town. Proving that in some cases the lack of verbosity is worse than being verbose, to begin with. Also because we now have a completely separate object handling transactions we can add on to it without worrying about cluttering any other classes in the process.

> Lastly, for a temporary town status change (e.g ruin, bankruptcy, or occupation) in town state, it seems very hard to try to do these via subclassing, so generally I have been keeping it simple and just keeping these somewhat generic flags in the Town object

Well, we already do it via subclassing, `isBankrupt()` isn't a `Town` or `Nation` method, it's a method within `BankAccount`. But regarding the infrastructure of making a new class for new methods, I would go ahead and say its worth it because once again, it will offer more flexibility if you decide to add more functionality to fallen towns in the future, you can add on to it as much as you like without having to worry about bloating the `Town` class. Especially if this aims to be "generic" I see more reason in subclasses rather than appending to already established classes.

This is just a recommendation, if you feel this implementation is optimum and works properly then we'll accept it.

Reply:
Hi,

- I believe the 'conquered' field (predating this PR) is correct.
- Similarly, I believe the nation 'neutral' field (predating this PR) is correct
- Similarly, I believe your upcoming update to the Account fields is also correct.

To demonstrate, I provide 3 example diagrams:

1. The below diagram demonstrates the OO aspects of your upcoming Account field updates.
![oo1](https://user-images.githubusercontent.com/50219223/86234169-702d7980-bb8e-11ea-877b-ebfafaffa669.png)

2. The below diagram demonstrate the OO aspects of the proposed subclassing of fields like 'conquered/peaceful/ruined'
![oo2](https://user-images.githubusercontent.com/50219223/86234204-7ae80e80-bb8e-11ea-9c15-4991f5d43aee.png)

3. The below diagram demonstrates what I believe is an appropriate way to reduce bloat in the Town object, while still doing the OO correctly.
![oo3](https://user-images.githubusercontent.com/50219223/86234220-820f1c80-bb8e-11ea-9f97-2a72c484cbe7.png)

Reply:
I'm fine with that implementation as well, perhaps instead of making the war status three separate fields, we can make it one field under one type.

Something akin to this:
```java
enum TownStatusType {
	PEACEFUL,
	FALLEN,
	CONQUERED
}
public class TownStatus {
	
	TownStatusType getStatusType() {
		...
	}
	
	int remainingStatusHours() {
		...
	}
	
	void setStatusType(TownStatusType type) {
		...
	}
	
}
```

Reply:
* On the data fields, ok cool. The real benefit of this would come not from updating the town class itself (becase a complexity cost would be paid, gaining only a small file readability increase in return), but rather from restructuring the db tables to match. 
This would reduce the db footprint when these features are Not used. Thus after the db update arrives, this can be looked at with the new structure - to see if it is feasible, and if the data saving benefit would actually be worth the complexity cost.

* On the enum idea, these are not mutually exclusive statuses (eg. A town could be peaceful and ruined, but not fallen). These are 3 separate and big features, with different sets of fields in each one e.g some features have 5 fields, some have two. Thus I would not advise adding a higher abstraction layer of this type.

* Also note that unlike peacefulness & invasion features, the town ruin feature (in this PR) is primarly, but not strictly, a war mechanic. Right now there are some servers using it without war. In fact there is an argument to move the configs out of the 'war common' section. But I wanted to get started on that common section, so I'll leave it there unless/until there is a request to move it.